### PR TITLE
Improve list_simplify() performance + errors

### LINKED ISSUE #935: Reconsider `can_simplify()` in `list_simplify()`
From 

```R
list_check_all_vectors(x, call = error_call)
can_simplify <- all(list_sizes(x) == 1L)
```

From https://github.com/tidyverse/purrr/pull/909#discussion_r968896274

### PR DESCRIPTION
**Pull request by hadley**
Fixes #935

### CODE REVIEW

**Comment by hadley** [comment]
Oooh, and name management from https://github.com/tidyverse/purrr/pull/894/files#r957388812.

**Review by DavisVaughan — APPROVED** [review]

File: R/list-simplify.R [hunk 1]
```diff
@@ -57,22 +63,34 @@ list_simplify_internal <- function(x,
...
   if (can_simplify) {
+
+    # TODO: use `error_call` when available
     tryCatch(
-      # TODO: use `error_call` when available
-      list_unchop(x, ptype = ptype),
+      {
+        names <- vec_names(x)
+        x <- vec_set_names(x, NULL)
+        out <- list_unchop(x, ptype = ptype)
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/list-simplify.R:92]: Oops yes. Let me refactor.

File: R/list-simplify.R [hunk 2]
```diff
@@ -83,8 +101,12 @@ simplify_impl <- function(x,
     )
   } else {
     if (strict) {
+      bad <- detect_index(x, function(x) vec_size(x) != 1)
```
[the remark this answers is not in the document]
Comment (hadley) [reply R/list-simplify.R]: I moved this error up into the earlier strict branch which I think makes it easier to follow.

**Review by DavisVaughan — APPROVED** [review]
Looks great now!

**Pull request merged**